File: process/OFM/functions_tidy.py
import pandas as pd
from pandas.api.types import is_string_dtype
import numpy as np

def tidy_ofm_apr(dataset, tabname):
    """
    This function will clean and transform a single sheet either post-censal or intercensal April 1 data from OFM. 
    Intercensal data has a different format than the postcensal (e.g. additional Place Code columns, column names, Jurisdiction string format) 
    """
    tabname = str.lower(tabname)
    tabname = str.replace(tabname, " ", '_')
    if tabname == 'population':
        tabname = 'total_population'
    if tabname == 'gq_population':
        tabname = 'group_quarters_population'
    if tabname == 'occupied_housing':
        tabname = 'occupied_housing_units'
    if tabname == 'total_housing':
        tabname = 'housing_units'

    # Intercensal specific munging
    if {'County Name', 'City Name'}.issubset(dataset.columns):
        dataset = dataset.rename(columns = {'County Name': 'County'})
        dataset['Jurisdiction'] = dataset['Jurisdiction'].str.replace('city', '')
        dataset['Jurisdiction'] = dataset['Jurisdiction'].str.replace('town','')
        dataset['Jurisdiction'] = dataset['Jurisdiction'].replace('\\s+', ' ', regex=True)
        dataset['Jurisdiction'] = dataset['Jurisdiction'].replace('\\s$', '', regex=True)
        drop_cols = ['City Name'] + list(dataset.columns[dataset.columns.str.contains('Code$')])
        dataset = dataset.drop(columns = drop_cols)

    # filter for region
    counties = ['King', 'Kitsap', 'Pierce', 'Snohomish']
    reg = dataset[dataset['County'].isin(counties)]
    
    # remove Line column, Filters 1 & 3
    reg = reg[~reg['Filter'].isin([1,3])]
    reg = reg.drop(columns = ['Line', 'Filter'])
    
    # pivot longer
    reg_pvt = reg.melt(id_vars=['County', 'Jurisdiction'], var_name='desc', value_name=tabname)
    reg_pvt['estimate_year'] = reg_pvt['desc'].str.extract('(^\\d+)')
    reg_pvt = reg_pvt.drop(columns = ['desc'])

    # rename and reorder
    reg_pvt.columns = map(str.lower, reg_pvt.columns)
    cols = ['county', 'jurisdiction', 'estimate_year', tabname]
    df = reg_pvt[cols]

    # convert to numeric cols
    if is_string_dtype(df[tabname]):
        df[tabname] = np.array(df[tabname], dtype=float)

    return(df)

File: process/OFM/functions_tidy.R
tidy_ofm_apr <- function(dataset, tabname) {

  # This function will clean and transform a single sheet either post-censal (population or housing) or intercensal April 1 data from OFM. 
  # Intercensal data has a different format than the postcensal (e.g. additional Place Code columns, column names, Jurisdiction string format) 

  d <- setDT(dataset)
  
  if('County.Name' %in% colnames(d) && 'City.Name' %in% colnames(d)) {
    dtype <- 'intercensal'
  } else if(length(str_subset(colnames(d), 'Two.or.More.Unit.Housing.Units')) > 1) {
    dtype <- 'housing'
  } else {
    dtype <- 'post'
  } 

  tabname <-  str_to_lower(tabname)
  tabname <-  str_replace_all(tabname, " ", '_')
  
  if(tabname == 'population') 'total_population'
  if(tabname == 'gq_population') 'group_quarters_population'
  if(tabname == 'occupied_housing') 'occupied_housing_units'
  if(tabname == 'total_housing') 'housing_units'

  # intercensal specific munging
  if('County.Name' %in% colnames(d) && 'City.Name' %in% colnames(d)) {
    setnames(d, 'County.Name', 'County')
    d <- d[, Jurisdiction := str_replace_all(Jurisdiction, 'city', '')]
    d <- d[, Jurisdiction := str_replace_all(Jurisdiction, 'town', '')]
    d <- d[, Jurisdiction := str_replace_all(Jurisdiction, '\\s+', ' ')]
    d <- d[, Jurisdiction := str_replace_all(Jurisdiction, '\\s$', '')]
    rm.cols <- c('City.Name', str_subset(colnames(d), '.*Code$'))
    d <- d[, !..rm.cols]
  }

  # filter for region
  counties <- c('King', 'Kitsap', 'Pierce', 'Snohomish')
  reg <- d[County %in% counties, ]
  
  # remove Line column, filter out 1 & 3
  reg <- reg[!(Filter %in% c(1, 3)), ][, !c('Line', 'Filter')]

  # pivot longer
  reg.pvt <- melt(reg, id.vars = c('County', 'Jurisdiction'), variable.name = 'desc', value.name = tabname)
  reg.pvt[, estimate_year := str_extract(desc, '^\\d+')]

  if(dtype == 'housing') {
    reg.pvt[, housing_type := str_extract(desc, '((?<=of\\.).*)')]
    reg.pvt[, housing_type := str_replace_all(housing_type, '\\.', '_')]
    reg.pvt[, housing_type := str_replace_all(housing_type, '\\W$', '')]
    reg.pvt[, housing_type := str_replace_all(housing_type, 'Moble', 'Mobile')]
    reg.pvt[, housing_type := str_to_lower(housing_type)]
  } 
  
  # munge
  reg.pvt[, desc := NULL]
  colnames(reg.pvt) <- str_to_lower(colnames(reg.pvt))
  
  if(dtype != 'housing') {
    # reorder
    cols <- c("county", "jurisdiction", "estimate_year", tabname)
    setcolorder(reg.pvt, cols)
    
  }

  # convert to numeric cols
  reg.pvt[, (tabname) := as.numeric(get(eval(tabname)))]
  
  if(dtype == 'housing') {
    reg.pvt <- dcast(reg.pvt, county + jurisdiction + estimate_year ~ housing_type, value.var = 'housing_units')
    cols <-  c('county', 'jurisdiction', 'estimate_year', 'total_housing_units', 'one_unit_housing_units', 'two_or_more_unit_housing_units', 'mobile_homes_and_specials')
    setcolorder(reg.pvt, cols)
  }
  
  return(reg.pvt)
}
